stl/inc/memory

```diff
@@ -1578,20 +1762,211 @@ private:
+        // nothing to do; _Storage._Value already destroyed by _Destroy()
+    }
+
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        _Maybe_rebind_alloc_t<_Alloc, _STD remove_all_extents_t<_Ty>> _Alty(this->_Get_val());
+        _Reverse_destroy_md_n_al(_Storage._Value, _STD extent_v<_Ty>, _Alty);
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        _Rebind_alloc_t<_Alloc, _Ref_count_bounded_array_alloc> _Al(this->_Get_val());
+        this->~_Ref_count_bounded_array_alloc();
+        _Deallocate_plain(_Al, this);
+    }
+};
+#endif // _HAS_CXX20
+
 // FUNCTION TEMPLATE make_shared
 template <class _Ty, class... _Types>
-_NODISCARD shared_ptr<_Ty> make_shared(_Types&&... _Args) { // make a shared_ptr
+#if _HAS_CXX20 && __cpp_concepts
+requires(!_STD is_array_v<_Ty>)
+#endif // _HAS_CXX20
+    _NODISCARD shared_ptr<_Ty> make_shared(_Types&&... _Args) { // make a shared_ptr to non-array object
     const auto _Rx = new _Ref_count_obj2<_Ty>(_STD forward<_Types>(_Args)...);
     shared_ptr<_Ty> _Ret;
     _Ret._Set_ptr_rep_and_enable_shared(_STD addressof(_Rx->_Storage._Value), _Rx);
     return _Ret;
 }
 
+#if _HAS_CXX20 && __cpp_concepts
+template <class _Ty>
+requires _STD is_unbounded_array_v<_Ty> _NODISCARD shared_ptr<_Ty> make_shared(
+    size_t _Count) { // make a shared_ptr to an unbounded array
+    using _Refc = _Ref_count_unbounded_array<_Ty>;
+    using _Storage =
+        _STD aligned_storage_t<(_STD max)(alignof(_Ty), alignof(_Refc)), (_STD max)(alignof(_Ty), alignof(_Refc))>;
+    constexpr size_t _Rx_size_div_by_storage_size = (_STD max)(sizeof(_Refc) / sizeof(_Storage), size_t(1));
+    constexpr double _Ty_size_div_by_storage_size =
+        static_cast<double>(sizeof(_STD remove_extent_t<_Ty>)) / static_cast<double>(sizeof(_Storage));
+    const auto _Rx = reinterpret_cast<_Refc*>(
+        new _Storage[static_cast<size_t>(ceil(_Count * _Ty_size_div_by_storage_size)) + _Rx_size_div_by_storage_size]);
+    new (_Rx) _Refc(_Count);
+    shared_ptr<_Ty> _Ret;
+    _Ret._Set_ptr_rep_and_enable_shared(_Rx->_Get_ptr(), _Rx);
+    return _Ret;
+}
+
+template <class _Ty>
+requires _STD is_unbounded_array_v<_Ty> _NODISCARD shared_ptr<_Ty> make_shared(size_t _Count,
+    const _STD remove_extent_t<_Ty>& _Val) { // make a shared_ptr to an unbounded array
+    using _Refc = _Ref_count_unbounded_array<_Ty>;
+    using _Storage =
+        _STD aligned_storage_t<(_STD max)(alignof(_Ty), alignof(_Refc)), (_STD max)(alignof(_Ty), alignof(_Refc))>;
+    constexpr size_t _Rx_size_div_by_storage_size = (_STD max)(sizeof(_Refc) / sizeof(_Storage), size_t(1));
+    constexpr double _Ty_size_div_by_storage_size =
+        static_cast<double>(sizeof(_STD remove_extent_t<_Ty>)) / static_cast<double>(sizeof(_Storage));
+    const auto _Rx = reinterpret_cast<_Refc*>(
+        new _Storage[static_cast<size_t>(ceil(_Count * _Ty_size_div_by_storage_size)) + _Rx_size_div_by_storage_size]);

```

Comment:
There are a lot of unchecked / undefended multiplications here. You need to protect any size calculation going into a memory allocation against integer overflow vulnerabilities.

Reply:
I believe that all multiplication and addition overflow is now checked by `_Calculate_bytes_for_flexible_array`.